File: src/main/java/com/notehub/notehub/service/note/NoteService.java
package com.notehub.notehub.service.note;

import java.util.Optional;
import java.util.UUID;

import org.springframework.data.domain.Page;

import com.notehub.notehub.entity.Note;
import com.notehub.notehub.entity.User;

public interface NoteService {
    Optional<Note> findById(UUID id);

    Optional<Note> findByTitle(String title);

    Page<Note> listNotes(int page, int size, String direction, String sortBy);

    Note createNote(Note note);

    Note updateNote(UUID id, Note note);

    void deleteNote(UUID id);

    int countUserNotes(User user);

    //List<Note> listUserNotes(User user); // todo

}

File: frontend/src/features/notes/services/NoteService.ts
import { api } from "../../../shared/util/api";
import { PaginationOptions, SearchType } from "../../search/types";
import type { NoteDetailDTO, NotesPageDTO, NoteSummaryDTO } from "../types";

export const NoteService = {
  async createNote(note: NoteSummaryDTO | NoteDetailDTO) {
    return api.post(`/notes`, note);
  },

  async getNotes(pagination: PaginationOptions) {
    const { page, size, direction, sortBy } = pagination;
    const response = await api.get<NotesPageDTO>(
      `/notes?page=${page}&size=${size}&direction=${direction}&sortBy=${sortBy}`
    );
    return response.data;
  },

  async getNoteByTitle(title: string) {
    const response = await api.get<NoteDetailDTO>(`/notes/title/${title}`);
    return response.data;
  },

  async searchNotes(
    term: string,
    searchType: SearchType,
    pagination: PaginationOptions
  ) {
    const { page, size, direction, sortBy } = pagination;
    const response = await api.get<NotesPageDTO>(
      `/notes/search?term=${term}&searchInContents=${Boolean(
        searchType === SearchType.CONTENT
      )}&page=${page}&size=${size}&direction=${direction}&sortBy=${sortBy}`
    );
    return response.data;
  },

  async updateNote(title: string, note: NoteSummaryDTO | NoteDetailDTO) {
    const response = await api.patch<NoteDetailDTO>(
      `/notes/title/${title}`,
      note
    );
    return response.data;
  },

  async deleteNote(title: string) {
    const response = await api.delete(`/notes/title/${title}`);
    return response.data;
  },
};
